### DISCUSSION TITLE
Route Network Optimization

### DISCUSSION THREAD

**Discussion by Smbat-A** [google/or-tools#2874] [Routing (and legacy CP) questions]
Hi,
I need to create Route where vehicle need to drive in all selected street

for Example we choose or get area by .shp file( I will get from file)
![maparea](https://user-images.githubusercontent.com/37016429/138830851-4e067c2e-1bf0-450d-9a88-5ba29763c3ac.jpg)

And need to get something like this

![maparearoute](https://user-images.githubusercontent.com/37016429/138831163-3a7292ab-2d25-4b31-be9a-485bfa264584.jpg)
  
Can I do this with Or-tools?

**Comment by Mizux** [comment]
Technically visiting all edge of a graph is the dual of visiting all nodes.

You can map for each street there is a node, there is an arc between two street if there are an intersection between them.
also you can use arc with `0` cost if street are connected, otherwise the shortest distance for traveling for one street to an other.

So yes it should be possible to apply the routing solver to this kind of problem but you'll need few works on your data to extract all street intersections etc...

**Reply by jmarca** [reply]
There are lots of references on the internet for edge covering problems. My humble and overlong contribution is here: 

https://activimetrics.com/blog/edge_covering/

In it I explain how I got from OSM data to a state where I could solve a street sweeping problem. 

I also did a webinar thing, focusing on the database side of the problem:

https://activimetrics.com/presentations/2020/pgconf/streetsweep/

**Reply by Smbat-A** [reply]
James can you explain me this path please, I don't understand where you get tables keep_way and new_way

Thank you. 

![mapGroup](https://user-images.githubusercontent.com/37016429/139670202-6c0ead21-1e1d-43c8-b23a-05edcf01a794.jpg)

**Reply by jmarca** [reply]
Honestly, if it is not obvious from what I wrote, then I probably deliberately skipped steps. 

I do this consulting thing for a living, so it doesn’t pay to give away *all* my tricks. If the reader can’t figure out the skipped steps on their own, then perhaps the reader might need to hire me. 

And if the reader is a student, then the gaps are homework exercises.
